pkg/controller/controller_binding_test.go

```diff
@@ -954,22 +991,77 @@ func TestReconcileServiceBindingDelete(t *testing.T) {
 	assertServiceBindingOrphanMitigationSet(t, updatedServiceBinding, false)
 
 	events := getRecordedEvents(testController)
-	assertNumEvents(t, events, 1)
 
 	expectedEvent := normalEventBuilder(successUnboundReason).msg("This binding was deleted successfully")
 	if err := checkEvents(events, expectedEvent.stringArr()); err != nil {
 		t.Fatal(err)
 	}
 }
 
+// TestReconcileServiceBindingDeleteUnresolvedClusterServiceClassReference
+// tests reconcileBinding to ensure a binding delete succeeds when a ClusterServiceClassRef
+// has not been resolved and no action has accrued for the binding.
+func TestReconcileServiceBindingDeleteUnresolvedClusterServiceClassReference(t *testing.T) {
+	_, fakeCatalogClient, fakeClusterServiceBrokerClient, testController, sharedInformers := newTestController(t, noFakeActions())
+
+	sharedInformers.ClusterServiceBrokers().Informer().GetStore().Add(getTestClusterServiceBroker())
+	sharedInformers.ClusterServiceClasses().Informer().GetStore().Add(getTestClusterServiceClass())
+	instance := &v1beta1.ServiceInstance{
+		ObjectMeta: metav1.ObjectMeta{Name: testServiceInstanceName, Namespace: testNamespace},
+		Spec: v1beta1.ServiceInstanceSpec{
+			PlanReference: v1beta1.PlanReference{
+				ClusterServiceClassExternalName: testNonExistentClusterServiceClassName,
+				ClusterServicePlanExternalName:  testClusterServicePlanName,
+			},
+			ExternalID: testServiceInstanceGUID,
+		},
+	}
+	sharedInformers.ServiceInstances().Informer().GetStore().Add(instance)
+	sharedInformers.ClusterServicePlans().Informer().GetStore().Add(getTestClusterServicePlan())
+
+	binding := &v1beta1.ServiceBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:              testServiceBindingName,
+			Namespace:         testNamespace,
+			DeletionTimestamp: &metav1.Time{},
+			Finalizers:        []string{v1beta1.FinalizerServiceCatalog},
+			Generation:        1,
+		},
+		Spec: v1beta1.ServiceBindingSpec{
+			ServiceInstanceRef: v1beta1.LocalObjectReference{Name: testServiceInstanceName},
+			ExternalID:         testServiceBindingGUID,
+		},
+		Status: v1beta1.ServiceBindingStatus{
+			UnbindStatus: v1beta1.ServiceBindingUnbindStatusRequired,
+		},
+	}
+
+	err := testController.reconcileServiceBinding(binding)
+	if err == nil {
+		t.Fatal("serviceclassref was nil + generation was 1: reconcile should abort the binding")
```

Comment:
What does "generation was 1" have to do with whether the reconciliation was successful or not? Also, I don't think the binding was aborted, was it? Or maybe I don't know what it means to "abort a binding".

Reply:
This was the test I added to test the case were you bind to a service instance with unresolved plan or class and you then try to delete the binding but can't. 

Binding Unbind status is going back to ServiceBindingUnbindStatusNotRequired, and there is no error expected. It should just delete.

Reply:
Ah, I see. I missed the part where it was testing a binding for which no BindRequest had been sent. My bad.